packages/client/lib/sync/fullsync.ts

```diff
@@ -3,17 +3,71 @@ import { Peer } from '../net/peer/peer'
 import { short } from '../util'
 import { Synchronizer, SynchronizerOptions } from './sync'
 import { BlockFetcher } from './fetcher'
+import VM from '@ethereumjs/vm'
+import { DefaultStateManager } from '@ethereumjs/vm/dist/state'
+import { SecureTrie as Trie } from '@ethereumjs/trie'
+const level = require('level')
 
 /**
  * Implements an ethereum full sync synchronizer
  * @memberof module:sync
  */
 export class FullSynchronizer extends Synchronizer {
   private blockFetcher: BlockFetcher | null
+  public vm: VM
 
   constructor(options: SynchronizerOptions) {
     super(options)
     this.blockFetcher = null
+
+    if (!this.config.vm) {
+      const db = level('./statedir')
+      const trie = new Trie(db)
+
+      const stateManager = new DefaultStateManager({
+        common: this.config.common,
+        trie,
+      })
+
+      this.vm = new VM({
+        common: this.config.common,
+        blockchain: this.chain.blockchain,
+        stateManager,
+      })
+    } else {
+      this.vm = this.config.vm
+      //@ts-ignore blockchain has readonly property
+      this.vm.blockchain = this.chain.blockchain
+    }
+
+    const self = this
+    this.chain.on('updated', function () {
+      // for some reason, if we use .on('updated', this.runBlocks), it runs in the context of the Chain and not in the FullSync context..?
+      // eslint-disable-next-line @typescript-eslint/no-floating-promises
+      self.runBlocks()
+    })
+    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+    this.chain.update()
```

Comment:
Why is it not possible on these two calls to just use `await`?

Reply:
`this.chain.update` is in the constructor, so we cannot use `await`. I did this to trigger `runBlocks`, but might also just call `this.runBlocks()` directly. It is for completeness: we do not wait until a new block arrives before we start executing the chain, this is especially useful if the node itself is offline, but it still has blocks left to process (and we just started the node).

The other one (`self.runBlocks`) can actually be awaited by converting it to an async function.